tchannel/container/heap.py

```diff
@@ -0,0 +1,86 @@
+from __future__ import absolute_import
+
+
+class HeapOperation(object):
+    def cmp(self, i, j):
+        """Compare the items in position i and j of the heap.
+
+        :param i: the first item's position of the heap list
+        :param j: the second item's position of the heap list
+        :return true if items[i] < items[j], otherwise false.
+        """
+        raise NotImplementedError()
+
+    def push(self, x):
+        """Push a new item into heap"""
+        raise NotImplementedError()
+
+    def pop(self):
+        """Pop an item from the heap"""
+        raise NotImplementedError()
+
+    def swap(self, i, j):
+        """swap items between position i and j of the heap"""
+        raise NotImplementedError()
+
+
+def init(h):
+    # heapify
+    n = len(h)
+    for i in range(n/2 - 1, -1):
+        down(h, i, n)
+
+
+def push(h, x):
+    h.push(x)
+    up(h, len(h)-1)
+
+
+def pop(h):
+    n = len(h) - 1
+    h.swap(0, n)
+    down(h, 0, n)
+    return h.pop()
+
+
+def remove(h, i):
+    n = len(h) - 1
+    if n != i:
+        h.swap(i, n)
+        down(h, i, n)
+        up(h, i)
+
+    return h.pop()
+
+
+def fix(h, i):
+    down(h, i, len(h))
+    up(h, i)
+
+
+def up(h, j):
+    while j > 0:
+        i = (j - 1) / 2
+        if i == j or not h.cmp(j, i):
```

Comment:
I was trying to make two points:

1. Given `j > 0`, it's impossible for `(j - 1) / 2 = j`. Just analytically, if you solve `j = (j - 1) / 2` for `j`, you get `j = -1`. The only time this condition will be hit is when `j = -1` and that's impossible because you have a `j > 0` condition on the while loop.
2. If by some magic `i == j`, then `h.cmp(i, j)` is always false. So, the second part of the `if` condition still holds.

Either way, the `i == j` check isn't required.